### PR TITLE
Increase coverage of leaflet.py 

### LINKED ISSUE #2600: Increasing coverage for leaflet.py
## Is your feature request related to a problem? ##
The coverage for analysis/leaflet.py is currently at ~ 69%. There should be some reasonably easy parts of the code which are missing coverage that could be tested. See: https://codecov.io/gh/MDAnalysis/mdanalysis/src/develop/package/MDAnalysis/analysis/leaflet.py

## Describe the solution you'd like ##
More tests, example cases that should be possible to cover include;
  - `_get_graph` without `self.pbc`. (lines 170-171)
  - `_get_graph` with `self.spare == False or True` (lines 173-183)
  - `_get_graph`, `self.spare == None` ValueError (lines 189-195)
  - using `update`
  - using `write_selection`

## Additional context ##
This is being raised as a feature request to point GSOC candidates towards code areas where increased coverage would be useful.

### LINKED ISSUE #2612: Missing selection module in leaflet.py
## Expected behavior ##

write_selection works

## Actual behavior ##

``` python
NameError: name 'selections' is not defined
``` 

## Code to reproduce the behavior ##

``` python
from MDAnalysis.analysis.leaflet import LeafletFinder, optimize_cutoff
from MDAnalysisTests.datafiles import Martini_membrane_gro

LIPID_HEAD_STRING = "name PO4"
universe = MDAnalysis.Universe(Martini_membrane_gro)
lipid_heads = universe.select_atoms(LIPID_HEAD_STRING)
lfls_ag = LeafletFinder(universe, lipid_heads, pbc=False)

lfls_ag.write_selection('test.vmd')
```

## Current version of MDAnalysis ##

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) '0.20.2-dev0'
- Which version of Python (`python -V`)? 3.7
- Which operating system? Ubuntu 18

### PR DESCRIPTION
**Pull request by yuxuanzhuang**
Fixes #2600
Fixes #2612

Changes made in this Pull Request:
- [x] _get_graph without self.pbc. (lines 170-171)
- [x] _get_graph with self.spare == False, True or None (lines 173-183)
- [x] _get_graph, self.spare == None ValueError (lines 189-195)
- [x] using update
- [x] using write_selection

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 1]
```diff
@@ -69,3 +79,19 @@ def test_optimize_cutoff(universe, lipid_heads):
...
+def test_cutoff_update(universe, lipid_heads):
+    lfls_ag = LeafletFinder(universe, lipid_heads, cutoff = 15.0, pbc=True)
+    lfls_ag.update(cutoff=1.0)
+    assert_almost_equal(lfls_ag.cutoff, 1.0, decimal=4)
+
+def test_write_selection(universe, lipid_heads, tmpdir):
+    lfls_ag = LeafletFinder(universe, lipid_heads, cutoff = 15.0, pbc=True)
+    with tmpdir.as_cwd():
+         filename = lfls_ag.write_selection('leaflet.vmd')
+         assert len(open('leaflet.vmd').readlines()) == 50, "Leaflet write selection not working"
```
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: Could you also test the inside of the file, i.e. say that the selection for say a vmd file would work here?

**Comment by yuxuanzhuang** [comment]
@IAlibay Thanks for the quick review! Just wondering what kind of test should be done for ´sparse´ option. As far as I can understand, they are just two different contact matrix data type. And I cannot possibly create a way larger matrix that will raise a ValueError from the pdb file for testing.

**Comment by orbeckst** [comment]
@IAlibay I am assigning you for right now but please feel free to un-assign yourself and/or ping someone else.

**Comment by IAlibay** [comment]
> @IAlibay Thanks for the quick review! Just wondering what kind of test should be done for ´sparse´ option. As far as I can understand, they are just two different contact matrix data type. And I cannot possibly create a way larger matrix that will raise a ValueError from the pdb file for testing.

@yuxuanzhuang Since the goal here is to increase coverage, at the very least what we should try is to test that the standard `True` and `False` options, when used, return the same values (you could possibly extend `test_leaflet_finder` with `pytest.mark.parametrize` (do let us know if you need pointers on this), to enumerate through each option and make sure the same outcome is captured every time.

With regards to the `ValueError`, I'm going to ping @orbeckst here.
For @orbeckst: @fiona-naughton and I looked into this more off-github, and we couldn't work out under what exact circumstances `contact_matrix`'s `returntype = numpy` would lead to this. Since this seems like code that pre-dates the Github days, we can't find the exact issue this relates to, so we thought you might have more insight here.

File: package/MDAnalysis/analysis/leaflet.py [hunk 2]
```diff
@@ -79,6 +79,7 @@
 
 from .. import core
 from . import distances
+from .. import selections
```
Comment (IAlibay) [inline package/MDAnalysis/analysis/leaflet.py]: Wow, really goes to show that tests are useful :) 

If you want, you could technically raise an issue about this and fix it within this PR, but either way, this should be recorded in the changelog.

Comment (yuxuanzhuang) [reply package/MDAnalysis/analysis/leaflet.py]: Thanks for your comments! I have done accordingly:)

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 3]
```diff
@@ -69,3 +72,98 @@ def test_optimize_cutoff(universe, lipid_heads):
...
+        4633 4645 4657 4669 }
+
+"""])
+
+        assert lines2one(open('leaflet.vmd').readlines()) == expected_output
+
+def test_component_index_is_not_none(universe, lipid_heads):
+    lfls_ag = LeafletFinder(universe, lipid_heads, cutoff = 15.0, pbc=True)
+    assert_almost_equal(len(lfls_ag.groups(component_index=0)), 180, decimal=4)
+
```
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: PEP8: no blank line at the end of the file.

File: testsuite/MDAnalysisTests/analysis/test_leaflet.py [hunk 4]
```diff
@@ -69,3 +72,98 @@ def test_optimize_cutoff(universe, lipid_heads):
     cutoff, N = optimize_cutoff(universe, lipid_heads, pbc=True)
     assert N == 2
     assert_almost_equal(cutoff, 10.5, decimal=4)
+
```
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_leaflet.py]: PEP8 change: you should have two blank lines between functions, same thing applies for functions further down.

**Comment by orbeckst** [comment]
@IAlibay can you squash & merge this PR? Great job reviewing it and moving it along.

**Comment by IAlibay** [comment]
@orbeckst for some reason, I don't appear to be authorized to merge the PR (as far as I can tell I have pretty much every other permission though, e.g. branch editing).

**Comment by richardjgowers** [comment]
@IAlibay try now?

**Comment by IAlibay** [comment]
Congrats on your first commit @yuxuanzhuang ! 🎉 

(cheers @richardjgowers that did the trick)

**Pull request merged**